### ISSUE TITLE
Mice switching IDs every few frames

### ISSUE THREAD

**Issue by ZelikowskyLab** [sgoldenlab/simba#277]
Hello,
I am having an issue where the ROI analysis is swapping the IDs of the mice every few frames.
Here is a GIF of the output video showing IDs switching:
![switchIDs](https://github.com/sgoldenlab/simba/assets/95374245/89a43c93-4b0f-4ee7-95fb-94c519645db5)

The IDs were not switching in sleap, so I assume this is a simba issue?
Here is a GIF of the sleap video not switching IDs:
![noSwitchIDs](https://github.com/sgoldenlab/simba/assets/95374245/7bd76fff-3a69-45e5-a80e-c5a7003fcb70)

System Info:
 - OS: Windows 10 Pro 22H2
 - Python Version: 3.8.5
 - Yes, using Anaconda

I have tried these fixes but none of them were successful so far:
- Outlier correction.
- Created a new project with optimized 8 body parts for each mouse.
- Created apparatus zone so mice are always in at least one zone.
- Interpolated sleap data by reconverting it in simba and using the “animal:nearest” interpolation option.
- Interpolated each body point.

Thanks for the help!

**Comment by sronilsson** [comment]
Hi @ZelikowskyLab !

SimBA to cause the issue, the columns values for some reason would have to be shuffled row-wise, while maintaining the the x and y order (the x and y for different body-parts are never mixed up). It might happen, but seems unlikely .

I had a look at the sleap created gif, the color never switches between animals, but there are a fair few missing data points and tracking oddities e.g. so I'm thinking that could be related somehow. 

<img width="764" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/d8d9dc13-14b1-4a22-a141-9ed9a6db8794">

Does sleap allow coloring all body-parts belonging to  animal in a separate palette,  rather than a uniform color per animal? 

If you use [THIS](https://github.com/sgoldenlab/simba/blob/master/docs/Tutorial_tools.md#visualize-pose-estimation-in-folder) tool to visualize the data in the `project_folder/csv/input_csv` directory after importing the data without smoothing and interpolation, do you still see the ID switches? Would help to pin down if its happening before or after data import.

**Comment by ZelikowskyLab** [comment]
Thanks for the response @sronilsson!

Yes, here is a GIF from SLEAP of all body parts being a different color:
![16_instances_color_nodes](https://github.com/sgoldenlab/simba/assets/95374245/13b0d4f2-7b31-4faf-89b4-757ec80eea3c)

I am currently working on manually assigning nodes to body parts for each frame of 5 seconds of the video in SLEAP, then am going to send this "perfect" data to Simba to try and test whether or not it is a Simba issue. I will post the result here when I am done with that.

**Comment by sronilsson** [comment]
Thanks @ZelikowskyLab! You might do this already but just a note, when you click to assign the animals to the different tracks as documented [here](https://github.com/sgoldenlab/simba/blob/master/docs/Multi_animal_pose.md#step-5-assigning-tracks-the-correct-identities), choose a frame when animals are clearly separated.

**Comment by ZelikowskyLab** [comment]
Thanks, yes we are choosing a frame where the animals are separated.
After I ran my manually assigned 5 second sleap data in simba, here were the results:

Manually assigned sleap video:
![sleap](https://github.com/sgoldenlab/simba/assets/95374245/f34403c5-f1c1-430b-a3f4-c82f80ae5d5a)

Result from simba:
![simba](https://github.com/sgoldenlab/simba/assets/95374245/f885c098-b58d-41f9-ac92-a82e3977a076)

All body points were present in the sleap video but the issue still persisted.
Thanks for the help

**Comment by sronilsson** [comment]
Hi @ZelikowskyLab thanks for digging into this. The error looks slightly different now and kind of worse with the bodypart predicts just fading away plus ID switches nearly every frame. The fastest way to solve it would probably be if there is s small reproducible example you can share and I can take a look? Do you have a small SimBA project with a short video together with the sleap data file you are importing that you could share on a GDrive?

**Comment by ZelikowskyLab** [comment]
Yes, we can upload the project files to a google drive. What email can we share it with?
Thanks

**Comment by sronilsson** [comment]
Can you share it with sronilsson@gmail.com ?

**Comment by ZelikowskyLab** [comment]
Yes, just shared it.
Please let me know if you have any questions about the files.

**Comment by sronilsson** [comment]
Thanks! Got it, and I can see the issue, it is happening as soon as the data has imported into SimBA, it must happen when SImBA is importing the slp file. Just for curiosity, have you tried to import h5 files from sleap into SimBA and do you see the same issues then?

**Comment by ZelikowskyLab** [comment]
Thanks for the observation!
I think there were four because it saved the two computer generated tracks as well as two manually edited tracks.
We deleted the manual tracks and still had the same issue with the mice switching IDs in Simba.
We also did a separate trial where we deleted the computer generated tracks in sleap, but we get the error: 
_SIMBA VALUE ERROR: The animal most proximal to click number 0 is animal named c57. The animal most proximal to click number 1 is also animal c57.Please indicate which animal is which using a video frame where the animals are clearly separated 🚨_
Therefore we can't test whether the dropped body points are contributing to the ID swapping.
I have attached all the files in a zip folder to this message. In the zip you will find a folder with the manual and computer track files. There is also a folder called 2mice8bpSept28b1 with the simba project that we used and a png of the image used for the user defined bp config.
[filesForGoldenLab.zip](https://github.com/sgoldenlab/simba/files/12754762/filesForGoldenLab.zip)

**Comment by sronilsson** [comment]
Thank you! let me test this out. In the meantime can you test one thing for me: 

I was given some code last week-ish from the SLEAP developers, to more rapidly import SLEAP .h5 tracking files, and I included this code in SimBA. If you haven't done so already, could you try how it looks in the latest version of SimBA by upgrading it with `pip install simba-uw-tf-dev --upgrade` and try to import your tracking data in .h5 format (and not .slp format) and let me know how it goes?

**Comment by ZelikowskyLab** [comment]
The update with the .h5 import worked!! Thank you.

**Comment by sronilsson** [comment]
Cool! Let me know if anything else comes up!
